some rewriting of the spec section on queries and transactions (#857)
eclipse-ee4j/jpa-api: spec/src/main/asciidoc/ch03-entity-operations.adoc

==== Synchronization to the Database [[a1955]]

Usually, a persistence context is synchronized to the database as
described below. However, as an exception, a persistence context of
type `SynchronizationType.UNSYNCHRONIZED`, or an application-managed
persistence context that has been created outside the scope of the
current transaction, is only synchronized to the database if it has
been explicitly joined to the current transaction by calling the
`joinTransaction()` method of `EntityManager`.

Assuming that the persistence context is joined to the current
transaction, the state of its managed entities is synchronized to the
database at transaction commit. This synchronization involves writing
to the database any pending modifications to the state of persistent

> entities and their relationships as specified above.

Reviewer: > specified above.

would've been nice if we can link it to a section 🫣

[NOTE]
====
It is particularly important to ensure that changes made to the inverse
side of a relationship are reflected by corresponding changes on the
owning side. The provider is not required to synchronize changes made
only at the inverse end of an association. Failure to modify the owning
side of a relationship typically results in the change being lost.
====

The persistence provider runtime is permitted to perform synchronization
to the database at other times, as long as a transaction is active and
the persistence context is joined to the transaction.

- The `flush()` method of `EntityManager` can be used by the application
   to force synchronization. When `flush()` is called, the provider must
   synchronize every pending modification made to every entity currently

>    associated with the persistence context to the database.

Reviewer 2: This might make the sentence easier to understand,
```suggestion
   synchronize to the database every pending modification made to
   every entity currently associated with the persistence context.
```

=== Query APIs [[a3125]]

The `Query` and `TypedQuery` interfaces are used to control parameter
binding and execution for both:

- static statements and queries declared in metadata, and
- dynamic statements and queries passed to methods of `EntityHandler`.

These interfaces also support pagination control and control over locking.

The `StoredProcedureQuery` interface is used to call stored procedures
defined in the database.

These interfaces may be found in <<persistence-api>>.

==== Query Execution [[query-execution]]

A query or statement written in the Jakarta Persistence query language
or in native SQL is executed via the `Query` or `TypedQuery` interface.

- A `SELECT` query specified in the Jakarta Persistence query language
  or using the Criteria API (as an instance of `CriteriaSelect`) is
  executed using one of the methods `getResultList()`, `getSingleResult()`,

>   or `getSingleResultOrNull()`.

Reviewer: should we mention `getResultStream` ? (and if so there's two more places right below)

See <<a2027>> for information about transaction-scoped and extended
persistence contexts. See <<a11431>> for further discussion of entity
manager use with no active transaction and persistence context types.

The `executeUpdate()` of `Query` or `TypedQuery` must not be called
without an active transaction contex.

Whether a `StoredProcedureQuery` can be invoked without an active
transaction context is determined by the transactional semantics and
the requirements of the stored procedure implementation and the database.
In particular, problems may occur if the stored procedure initiates a
transaction and a transaction is already active. The state of entities
returned by a stored procedure query invocation is determined according
to the rules for `Query` or `TypedQuery` as described above.

If the persistence context is joined to the transaction, any runtime
exception thrown by any method not explicitly listed below of `Query`,
`TypedQuery`, or `StoredProcedureQuery` causes the current transaction
to be marked for rollback, unless the exception is an instance of one

> of the following exception types:

Author: Frankly I dislike both options ~ equally.

- `NonUniqueResultException` or `NoResultException`, or
- `QueryTimeoutException` or `LockTimeoutException`.

On any database platform on which a query timeout causes an immediate
transaction rollback, the persistence provider must throw a
`PersistenceException` instead of a `QueryTimeoutException`.

Certain methods never cause the current transaction to be marked for
rollback:

- a runtime exception thrown by the following methods of the `Query`,
  `TypedQuery`, and `StoredProcedureQuery` interfaces do not cause the
  current transaction to be marked for rollback:
  `getLockMode`, `getParameters`, `getParameter`, `getParameterValue`,

>   `getOutputParameterValue`, and

Reviewer: ```suggestion
Runtime exception thrown by certain methods never cause the current 
transaction to be marked for rollback:

- `getLockMode`, `getParameters`, `getParameter`, `getParameterValue`,
  `getOutputParameterValue` methods of the `Query`,
  `TypedQuery`, and `StoredProcedureQuery` interfaces, and

- any method of the `Tuple`, `TupleElement`, or `Parameter` interface.
```

maybe something along the lines ^ so that we don't repat the "transaction to be marked for rollback" so many times 🫣

Named parameters are case-sensitive. The mixing of named and positional
parameters is invalid.

> A named parameter of a Jakarta Persistence query language query is an

Reviewer: ```suggestion
A named parameter of a JPQL query is an
```

maybe ? so that we don't say query query 🫣
